be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	/**
+	 * Iterate over the log records from the begining and check whether all
+	 * the records preceeding this are persistent. If not, we cannot prune
+	 * the record with the given id.
+	 */
+
+	m0_tl_for (lrec, log->u.dl_tlist, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
+					      M0_DTPS_PERSISTENT))
+			return M0_ERR(-EPROTO);
+
+		rc = m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id, id);
+		if (rc != M0_DTS_LT)
+			break;
+	} m0_tl_endfor;
+
+	if (rc != M0_DTS_EQ)
+		return -ENOENT;
+
+	/* rec is a pointer to the record matching the input id. Delete all the
+	 * previous records and then this record. */
+	while ((currec = lrec_tlist_pop(log->u.dl_tlist)) != rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(currec));
+		be_dtm0_log_rec_fini(&currec, tx);
+	}
+
+	be_dtm0_log_rec_fini(&rec, tx);
+	return rc;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_clear(struct m0_be_dtm0_log *log)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	/* This function is expected to be called only on the client side where
+	 * the log will always be a volatile log.
+	 * TBD: can we change the name of dl_is_persistent to something more
+	 * intutive.
+	 */
+	M0_ASSERT(!log->dl_is_persistent);
+
+	m0_tl_teardown(lrec, log->u.dl_tlist, rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(rec));
+		M0_ASSERT(m0_dtm0_tx_desc_state_eq(&rec->dlr_dtx.dd_txd,
+						   M0_DTPS_PERSISTENT));
+		be_dtm0_log_rec_fini(&rec, NULL);
+	}
+	M0_POST(lrec_tlist_is_empty(log->u.dl_tlist));
+}
+
+M0_INTERNAL int m0_be_dtm0_volatile_log_insert(struct m0_be_dtm0_log  *log,
+					       struct m0_dtm0_log_rec *rec)
+{
+	int rc;
+
+	M0_ENTRY();
+	/* TODO: dissolve dlr_txd and remove this code */
+	rc = m0_dtm0_tx_desc_copy(&rec->dlr_dtx.dd_txd, &rec->dlr_txd);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	lrec_tlink_init_at_tail(rec, log->u.dl_tlist);
+	return M0_RC(rc);
+}
+
+M0_INTERNAL void m0_be_dtm0_volatile_log_update(struct m0_be_dtm0_log  *log,
+						struct m0_dtm0_log_rec *rec)
+{
+	/* TODO: dissolve dlr_txd and remove this code */
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, &rec->dlr_dtx.dd_txd);
+}
+
+M0_INTERNAL bool m0_be_dtm0_plog_can_prune(struct m0_be_dtm0_log    *log,
+					   const struct m0_dtm0_tid *id,
+					   struct m0_be_tx_credit   *accum)
+{
+	/* This assignment is meaningful as it covers the empty log case */
+	int                     rc = M0_DTS_LT;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_be_list      *dl_list = log->u.dl_list;
+	struct m0_be_tx_credit  cred = M0_BE_TX_CREDIT(0, 0);
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_be_list_for(lrec, dl_list, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd, M0_DTPS_PERSISTENT))
+			return false;
+
+		if (accum != NULL) {
+			m0_be_dtm0_log_credit(M0_DTML_PRUNE, NULL, NULL, log->dl_seg, rec, &cred);
+		}
```

Comment:
No parentheses are needed for a single line of clause.

And check if this line exceeds 80 chars.

Reply:
Done. Corrected!